tools/SampleAnalyzer/Commons/Base/SFSTaggerBase.cpp

```diff
@@ -0,0 +1,508 @@
+
+// SampleAnalyser headers
+#include "SampleAnalyzer/Commons/Base/SFSTaggerBase.h"
+#include "SampleAnalyzer/Commons/Service/RandomService.h"
+#include "SampleAnalyzer/Commons/Service/Physics.h"
+#include "SampleAnalyzer/Commons/Service/PDGService.h"
+
+namespace MA5 {
+
+    /// Print parameters
+    void SFSTaggerBase::PrintParam() const
+    {
+        /// Print B-taggging options
+        std::string excl = _options.btag_exclusive ? "Exclusive" : "Inclusive";
+        INFO << "        with bjet: matching ΔR = "
+             << _options.btag_matching_deltaR
+             << " ; " << excl << endmsg;
+        excl = _options.ctag_exclusive ? "Exclusive" : "Inclusive";
+
+        /// Print C-tagging options
+        if (_options.enable_ctagging) {
+            INFO << "        with cjet: matching ΔR = "
+                 << _options.ctag_matching_deltaR
+                 << " ; " << excl << endmsg;
+        }
+
+        /// Print Tau-tagging options
+        if (_options.tautag_jetbased) {
+            excl = _options.ctag_exclusive ? "Exclusive" : "Inclusive";
+            INFO << "        with tau : matching ΔR = "
+                 << _options.tautag_matching_deltaR
+                 << " ; " << excl << endmsg;
+        } else {
+            INFO << "        with tau : hadron-based tagging" << endmsg;
+        }
+    }
+
+    /// Convert jet to tau
+    MAfloat32 SFSTaggerBase::tau_tagging_eff(const RecJetFormat &jet, TaggerStatus status) const
+    {
+        RecTauFormat myTau;
+        myTau.setMomentum(jet.momentum());
+        myTau.ntracks_   = jet.ntracks();
+        myTau.mc_        = jet.mc_;
+
+        if (status == TaggerStatus::MID) return mid_tau_tagging_eff(myTau);
+        else if (status == TaggerStatus::TIGHT) return tight_tau_tagging_eff(myTau);
+        else return loose_tau_tagging_eff(myTau);
+    }
+
+    /// Execute tagger
+    void SFSTaggerBase::Execute(EventFormat &myEvent) const
+    {
+        /// Shortcut for global event variables
+        MAfloat64 & THT  = myEvent.rec()->THT();
+        MAfloat64 & Meff = myEvent.rec()->Meff();
+
+        /// Truth level B/C tagging
+        BJetTagging(myEvent);
+        CJetTagging(myEvent);
+
+        /// Run tau tagging
+        if (_options.tautag_jetbased) JetBasedTauTagging(myEvent);
+
+        /// Get initial number of taus.
+        /// @attention if tau tagging is jet based Ntau will be zero
+        MAuint32 Ntau = myEvent.rec()->taus().size();
+        std::vector<MAuint32> toRemove;
+        /// Jet tagging with detector effects
+        if (_isJetTaggingOn)
+        {
+            MAint32 ijet = -1;
+            for (auto &jet: myEvent.rec()->jets())
+            {
+                ijet++;
+                /// We have a true b-jet: is it b-tagged?
+                if (jet.true_btag())
+                {
+                    if (RANDOM->flat() > loose_b_tagging_eff(jet)) jet.setLooseBtag(false);
```

Comment:
Here. we should use the same random number. Otherwise, a jet tagged as a medium b-jet has a chance not to be light tagged too. I have changed quite a lot of minor things here. Please double check you are happy with them.

Reply:
Could you please elaborate on this? I don't think I completely understand what you mean. Loose/tight/med are entirely independent of each other.